examples/ota_http_client.rs

```diff
@@ -0,0 +1,147 @@
+//! OTA update example.
+//!
+//! In this example, we are checking for updates from a distant server hosting the latest firmware.
+//! The hosting server is supposed to check if an update should be done or not by responding with
+//! a 200 (OK, you should update) or 304 (Not Modified, you are up to date) status code.
+//!
+//! For this example to work, you need a OTA ready partition table.
+//! You need at least 2 OTA partitions (see https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/kconfig.html#config-partition-table-type)
+//!
+//! The most common starting point for OTA is the "Factory app, two OTA definitions" layout (https://github.com/espressif/esp-idf/blob/master/components/partition_table/partitions_two_ota.csv).
+//!
+//! To use a custom partition table, download the CSV file of your choice (or create a custom one)
+//! and use either use the `--partition-table` option of `espflash`, or set this option in your `espflash.toml` file.
+//!
+//! After a successful OTA update, you will need to reset the `otadata` partition. Otherwise, the ESP
+//! will continue to boot on the second partition, while `cargo run` is flashing the first one by default.
+//! To reset the `otadata` partition, add `--erase-parts otadata` to the runner command in `.cargo/config.toml`.
+
+use anyhow::Context;
+use embedded_svc::http::client::Client as HttpClient;
+use esp_idf_svc::http::client::{EspHttpConnection, Method};
+use esp_idf_svc::io;
+use esp_idf_svc::log::EspLogger;
+use esp_idf_svc::wifi::{BlockingWifi, ClientConfiguration, Configuration, EspWifi};
+use esp_idf_svc::{eventloop::EspSystemEventLoop, nvs::EspDefaultNvsPartition};
+use esp_idf_svc::{
+    hal::peripherals::Peripherals,
+    ota::{EspOta, SlotState},
+};
+
+use log::{error, info};
+
+const VERSION: &str = "1.0.0"; // You can pull this from an environment variable at build time using env! macro.
+const OTA_FIRMWARE_URI: &str = "http://your.domain/path/to/firmware";
+
+const SSID: &str = env!("WIFI_SSID");
+const PASSWORD: &str = env!("WIFI_PASS");
+
+mod http_status {
+    pub const OK: u16 = 200;
+    pub const NOT_MODIFIED: u16 = 304;
+}
+
+fn main() -> anyhow::Result<()> {
+    esp_idf_svc::sys::link_patches();
+    EspLogger::initialize_default();
+
+    // Setup Wifi
+
+    let peripherals = Peripherals::take()?;
+    let sys_loop = EspSystemEventLoop::take()?;
+    let nvs = EspDefaultNvsPartition::take()?;
+
+    let mut wifi = BlockingWifi::wrap(
+        EspWifi::new(peripherals.modem, sys_loop.clone(), Some(nvs))?,
+        sys_loop,
+    )?;
+
+    connect_wifi(&mut wifi)?;
+
+    let mut client = HttpClient::wrap(EspHttpConnection::new(&Default::default())?);
+
+    check_for_updates(&mut client)?;
+
+    check_valid_state();
```

Comment:
This has to be done post app-reboot, once the newly-updated partition was booted. Perhaps document this?

Reply:
In the body of the function, the check happens only if the currently running slot is not yet marked as valid.

I found it easier like this, to avoid having complex OTA checks in the middle of the code ^^' In my project for example, I call this function right after I've successfully opened my websocket to the server.

Reply:
I think you are missing the point as to what the check is supposed to do.
Read this: https://docs.espressif.com/projects/esp-idf/en/v4.2/esp32/api-reference/system/ota.html#app-rollback

My point is, **unless you reboot** you won't be running the _new_ firmware. And then, unless you are running the _new_ firmware, what exactly are you checking? `check_valid_state` is supposed to be invoked once the _new_ firmware is running for - say - a couple of minutes and did not crash in the meantime... It is completely pointless to tell the bootloader "hey, the new firmware is fine!" given that you are actually not yet running the new firmware in the first place.

Reply:
The `check_for_updates` function is rebooting the board once the update is done :-) 

So if the program is reaching `check_valid_state()` call, it means we are already up to date. So we should verify if the running slot is valid or in pending verification state, which is what `check_valid_state` is doing.

I agree in a normal project, you would probably mark the check later, once you are sure you are able to run normally.

I've added some comments to try to make it more clear.

Reply:
Let's add information about that too :-)
I didn't noticed, since my app is already too large to fit 1 FR + 2 OTA partitions in my 4MB flash ^^'